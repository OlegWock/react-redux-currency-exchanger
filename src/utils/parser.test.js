import parse from './parser';

describe('Parser should validate input correcrtly', () => {
    it('should parse simple expressions', () => {
        let result1 = parse('1+4EUR*0.8', 'USD');
        expect(result1.valid).toBe(true);
        expect(JSON.stringify(result1.tokens)).toMatchSnapshot();
    });

    it('should correctly parse complex expressions', () => {
        let {valid, tokens} = parse('(13$+4.5EUR - 8GBP) * 0.8 / 2');
        expect(valid).toBe(true);

        expect(JSON.stringify(tokens)).toMatchSnapshot();
    });

    it('should correctly parse invalid expressions', () => {
        let {valid} = parse('))(()(24()');
        expect(valid).toBe(false);

        valid = parse('))sdfs').valid;
        expect(valid).toBe(false);

        valid = parse('()*()').valid;
        expect(valid).toBe(false);

        valid = parse('-1EUR+3#USD').valid;
        expect(valid).toBe(false);


    })
});