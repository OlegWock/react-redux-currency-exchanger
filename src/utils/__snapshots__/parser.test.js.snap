// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser should validate input correcrlt should correctly parse complex expressions 1`] = `"[{\\"type\\":\\"bracket\\",\\"value\\":\\"(\\"},{\\"type\\":\\"number\\",\\"value\\":\\"13\\",\\"currency\\":\\"USD\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"+\\"},{\\"type\\":\\"number\\",\\"value\\":\\"4.5\\",\\"currency\\":\\"EUR\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"-\\"},{\\"type\\":\\"number\\",\\"value\\":\\"8\\",\\"currency\\":\\"GBP\\"},{\\"type\\":\\"bracket\\",\\"value\\":\\")\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"*\\"},{\\"type\\":\\"number\\",\\"value\\":\\"0.8\\",\\"currency\\":\\"USD\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"/\\"},{\\"type\\":\\"number\\",\\"value\\":\\"2\\",\\"currency\\":\\"USD\\"}]"`;

exports[`Parser should validate input correcrlt should parse simple expressions 1`] = `"[{\\"type\\":\\"number\\",\\"value\\":\\"1\\",\\"currency\\":\\"USD\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"+\\"},{\\"type\\":\\"number\\",\\"value\\":\\"4\\",\\"currency\\":\\"EUR\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"*\\"},{\\"type\\":\\"number\\",\\"value\\":\\"0.8\\",\\"currency\\":\\"USD\\"}]"`;

exports[`Parser should validate input correcrtly should correctly parse complex expressions 1`] = `"[{\\"type\\":\\"bracket\\",\\"value\\":\\"(\\"},{\\"type\\":\\"number\\",\\"value\\":\\"13\\",\\"currency\\":\\"USD\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"+\\"},{\\"type\\":\\"number\\",\\"value\\":\\"4.5\\",\\"currency\\":\\"EUR\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"-\\"},{\\"type\\":\\"number\\",\\"value\\":\\"8\\",\\"currency\\":\\"GBP\\"},{\\"type\\":\\"bracket\\",\\"value\\":\\")\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"*\\"},{\\"type\\":\\"number\\",\\"value\\":\\"0.8\\",\\"currency\\":\\"USD\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"/\\"},{\\"type\\":\\"number\\",\\"value\\":\\"2\\",\\"currency\\":\\"USD\\"}]"`;

exports[`Parser should validate input correcrtly should parse simple expressions 1`] = `"[{\\"type\\":\\"number\\",\\"value\\":\\"1\\",\\"currency\\":\\"USD\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"+\\"},{\\"type\\":\\"number\\",\\"value\\":\\"4\\",\\"currency\\":\\"EUR\\"},{\\"type\\":\\"operator\\",\\"value\\":\\"*\\"},{\\"type\\":\\"number\\",\\"value\\":\\"0.8\\",\\"currency\\":\\"USD\\"}]"`;
